core/async/src/test_loop/event_handler.rs

```diff
@@ -0,0 +1,135 @@
+        IndexedLoopEventHandler::new(Box::new(self), index)
+    }
+}
+
+/// A convenient trait to TryInto, or else return the original object. It's useful
+/// for implementing event handlers.
+pub trait TryIntoOrSelf<R>: Sized {
+    fn try_into_or_self(self) -> Result<R, Self>;
+}
+
+impl<R, T: TryInto<R, Error = T>> TryIntoOrSelf<R> for T {
+    fn try_into_or_self(self) -> Result<R, Self> {
+        self.try_into()
+    }
+}
+
+/// Implements .widen() for an event handler.
+pub struct WideningEventHandler<Data, Event> {
+    inner: Box<dyn LoopEventHandler<Data, Event>>,
+}
+
+impl<
+        Data,
+        Event,
+        OuterData: AsMut<Data>,
+        OuterEvent: TryIntoOrSelf<Event> + From<Event> + 'static,
+    > LoopEventHandler<OuterData, OuterEvent> for WideningEventHandler<Data, Event>
+{
+    fn init(&mut self, sender: DelaySender<OuterEvent>) {
+        self.inner.init(sender.narrow())
+    }
+
+    fn handle(&mut self, event: OuterEvent, data: &mut OuterData) -> Option<OuterEvent> {
+        let mut inner_data = data.as_mut();
+        match event.try_into_or_self() {
+            Ok(e) => self.inner.handle(e, &mut inner_data).map(|e| e.into()),
+            Err(e) => Some(e),
+        }
+    }
+}
+
+/// Implements capture_events().
+struct CaptureEvents;
+
+impl<Event> LoopEventHandler<Vec<Event>, Event> for CaptureEvents {
+    fn handle(&mut self, event: Event, data: &mut Vec<Event>) -> Option<Event> {
+        data.push(event);
+        None
+    }
+}
+
+/// An event handler that puts the event into a vector in the Data, as long as
+/// the Data contains a Vec<CapturedEvent>. (Use widen() right after).
+///
+/// This is used on output events so that after the test loop finishes running
+/// we can assert on those events.
+pub fn capture_events<Event>() -> impl LoopEventHandler<Vec<Event>, Event> {
+    CaptureEvents
+}
+
+/// Implements periodic_timer().
+struct PeriodicTimer<Data, Event: Clone + PartialEq> {
+    interval: Duration,
+    event: Event,
+    delay_sender: Option<DelaySender<Event>>,
+    func: Box<dyn Fn(&mut Data) + 'static>,
+}
+
+impl<Data, Event: Clone + PartialEq> LoopEventHandler<Data, Event> for PeriodicTimer<Data, Event> {
+    fn init(&mut self, sender: DelaySender<Event>) {
+        self.delay_sender = Some(sender);
+        self.delay_sender.as_ref().unwrap().send_with_delay(self.event.clone(), self.interval);
+    }
+
+    fn handle(&mut self, event: Event, data: &mut Data) -> Option<Event> {
+        if event == self.event {
+            (self.func)(data);
+            self.delay_sender.as_ref().unwrap().send_with_delay(self.event.clone(), self.interval);
+        }
+        None
+    }
+}
+
+/// Periodically sends to the event loop the given event by the given interval.
+/// Each time this event is handled, the given function is called.
+/// The first invocation is triggered after the interval, not immediately.
+pub fn periodic_timer<Data, Event: Clone + PartialEq>(
```

Comment:
In Rust it is called [Interval](https://docs.rs/tokio/latest/tokio/time/fn.interval.html), in Golang it is called [Ticker](https://pkg.go.dev/time#Ticker). I'd recommend to stick to one of the existing conventions.

Reply:
Sounds good.